<a href="https://colab.research.google.com/github/pavan-potnuru/hotel-booking-analysis-team-repository/blob/main/Copy_of_Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

In [1]:
# Importing the libraries
import numpy as np
import pandas as pd
#Mounting the drive
from google.colab import drive
drive.mount('/content/drive')


In [2]:
#Mounting the drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Importing the dataset
dataset = pd.read_csv('/content/drive/MyDrive/Copy of Hotel Bookings.csv')

In [4]:
dataset.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [6]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
dataset.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

In [28]:
dataset.value_counts("hotel")

hotel
City Hotel      79330
Resort Hotel    40060
dtype: int64

**Dataset Information:**

hotel: Type of the hotel (Categorical)

is_canceled: Whether the booking is cancelled 0-NotCanceled, 1-Canceled (Categorical)

lead_time: No of days prior to which the booking is made (Numeric)

arrival_date_year: Year of arrival to hotel (Numeric)

arrival_date_month: Month of arrival (Categorical)

arrival_date_week_number: Week of the year of arrival (Numeric)

arrival_date_day_of_month: Day of the Month of arrival (Numeric)

stays_in_weekend_nights: No of weekend nights during the stay (Numeric)

stays_in_week_nights: No of week nights during the stay (Numeric)

adults: No of adults (Numeric)

children: No of Children (Numeric)

babies: No of babies (Numeric)

meal: Type of meal opted (Categorical)

country: Country of the customer (Categorical)

market_segment: means through which booking happend (Categorical)

distribution_channel: means through which booking happend (Categorical)

is_repeated_guest: Whether the customer is repeated 0-New Customer, 1-Repeated Customer (Categorical)

previous_cancellations: No of cancellations previously done by customer (Numeric)

previous_bookings_not_canceled: No of bookings completed without cancellation in the past by the customer (Numeric)

reserved_room_type: Type of room reserved (Categorical)

assigned_room_type: Type of room assigned (Categorical)

booking_changes: No of changes made after booking (Numeric)

deposit_type: Type of deposit (Categorical)

agent: Agent code (Categorical)

company: Company code (Categorical)

days_in_waiting_list: No of days in waiting list (Numeric)

customer_type: Type of customer (Categorical)

adr: Average Daily rate (Numeric)

required_car_parking_spaces: No of cars to be parked (Numeric)

total_of_special_requests: No of special requests made by customer (Numeric)

reservation_status: Status of booking (Categorical)

reservation_status_date: Date at which the last status was set (Date)

In [29]:
#Copying data set
df=dataset.copy()

# **Data Cleaning**

Data cleaning is a crucial part of EDA. In data cleaning we perform operations like:

1. Removing duplicates

2. Handling missing values

3. Handling features (i.e., Merging similar features, removing irrelavent columns and correcting the data types)

# 1. **Removing Duplicates**

In [30]:
#show duplicate count
df[df.duplicated()].shape

(31994, 32)

In [32]:
#Dropping duplicates
df.drop_duplicates(inplace=True)
#confirming duplicates are dropped
df[df.duplicated()].shape

(0, 32)

In [34]:
#no of observations after duplicates are removed
df.shape

(87396, 32)

# **2. Handling missing values**

In [37]:
#Null values count in each column
df.isnull().sum()

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_year                     0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              4
babies                                0
meal                                  0
country                             452
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             12193
company                           82137


From above we can see that only 4 columns namely Agent, Company, Country and Children have null values. Let us see how we can handle them

Agent and Company have their codes as values. There maybe a chance that the customer has done booking neither through agent nor through a company. In this case the values may be null, so we can assign 0 as a code for such customers.

In [38]:
#filling NaN values with 0
df[['company','agent']]=df[['company','agent']].fillna(0)

Country values with null values can be considered as other category, and replace NaN values with "Others"

In [39]:
#filling NaN values with "Others"
df['country']=df['country'].fillna('Others')

There are only 4 children values with NaN, they can be replaced with 0, as there is a high chance that people with 0 children may be missing to fill the column

In [41]:
#Filling NaN values with '0'
df['children']=df['children'].fillna(0)

In [42]:
#Confirming there are no null values left
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

There may be a chance that Adults+Children+babies is 0 for some rows. i.e, No persons staying in the rooms. We can remove such rows

In [43]:
#Checking for values where sum of adults, babies and children are 0
df[df['adults']+df['babies']+df['children'] == 0].shape

(166, 32)

There are 166 such rows. Let us remove such rows

In [46]:
#Dropping values where the sum is 0 for the adults, children and babies
df.drop(df[df['adults']+df['babies']+df['children'] == 0].index, inplace = True)

# **3. Handling Features**

In [51]:
df[['market_segment','distribution_channel']].value_counts()

market_segment  distribution_channel
Online TA       TA/TO                   51254
Offline TA/TO   TA/TO                   13700
Direct          Direct                  11467
Corporate       Corporate                3893
Groups          TA/TO                    3606
                Corporate                 665
                Direct                    651
Complementary   Direct                    539
Direct          TA/TO                     229
Aviation        Corporate                 216
Corporate       TA/TO                     154
                Direct                    153
Online TA       GDS                       136
                Direct                    128
Offline TA/TO   Corporate                  95
Direct          Corporate                  81
Complementary   Corporate                  78
                TA/TO                      75
Offline TA/TO   GDS                        44
Online TA       Corporate                  34
Offline TA/TO   Direct                     

From the above we can see that even though market_segment and distribution_channel seems to have similar values, there are sub categories in each market_segment. So we cannot drop any of the column

We can generate another column by adding up stays_in_weekend_nights and stays_in_week_nights as Total_nights_stayed



In [52]:
# Adding total_nights_stayed column
df['total_nights_stayed'] = df['stays_in_weekend_nights']+df['stays_in_week_nights']

We can also generate another column by adding adults, children and babies as total_people

In [53]:
#adding total_people
df['total_people'] = df['adults']+df['children']+df['babies']

In [54]:
#checking the num of columns after adding new columns
df.shape

(87230, 34)

Changing to appropriate data types

In [58]:
#Converting datatype of columns 'children', 'company', 'agent','total_people'and 'total_nights_stayed' from float to int.
df[['children', 'company', 'agent','total_people','total_nights_stayed']] = df[['children', 'company', 'agent','total_people','total_nights_stayed']].astype('int64')

In [57]:
# changing datatype of column 'reservation_status_date' to date_type.
df['reservation_status_date'] = pd.to_datetime(df['reservation_status_date'], format = '%Y-%m-%d')